### PR TITLE
RFC: Keras Mixed Precision API

### PR DESCRIPTION
**Pull request by reedwm**
This RFC will be open for comment until October 14, 2020.

| Status        | Proposed                                             |
:-------------- |:---------------------------------------------------- |
| **RFC #**     | [293](https://github.com/tensorflow/community/pull/293) |
| **Author(s)** | Reed Wanderman-Milne (reedwm@google.com              |
| **Sponsor**   | Francois Chollet (fchollet@google.com)               |
| **Updated**   | 2020-09-29                                           |

# Objective

Make mixed precision easy to use in Keras.

### CODE REVIEW

File: rfcs/20200929-keras-mixed-precision.md [hunk 1]
```diff
@@ -0,0 +1,1152 @@
...
+```python
+layer = tf.keras.layers.Dense()
+print(layer.dtype_policy.name)  # float32, the default dtype policy
+tf.keras.mixed_precision.set_global_policy("mixed_float16")
+layer2 = tf.keras.layers.Dense()
+print(layer2.dtype_policy.name)  # mixed_float16
+print(layer.dtype_policy.name)  # float32. A layer dtype policy never changes.
+```
+
+The global policy is only used by Keras to determine the default layer policy, and has no other purpose. The next section describes in detail how a layer users dtype policies.
```
Comment (yimeisun123) [inline rfcs/20200929-keras-mixed-precision.md:167]: typo, "... layer **users** dtype ...", "uses"

File: rfcs/20200929-keras-mixed-precision.md [hunk 2]
```diff
@@ -0,0 +1,1152 @@
...
+  model = create_model(num_layers=10)
+  model.compile(...)
+  dataset = ...
+  model.fit(dataset)
+
+if __name__ == "__main__":
+  app.run(main)
+```
+
+Alternatively, the global policy can be set in `create_model`. This will work in cases where multiple models are not built in parallel using threads, and where every time a model is created, the global policy is set to an appropriate value.
```
Comment (yimeisun123) [inline rfcs/20200929-keras-mixed-precision.md:252]: I am a little confused about the case here. This is the way for use case "where multiple models are **not** built in parallel using threads" or "where multiple models are built in parallel using threads"?

**Review by di-d6 — COMMENTED** [review]
[rfcs/20200929-keras-mixed-precision.md]()

File: rfcs/20200929-keras-mixed-precision.md (hunk 2 above)
Comment (reedwm) [reply rfcs/20200929-keras-mixed-precision.md]: I removed the part about threads and reworded this paragraph. I was trying to state that you cannot create a float32 and a mixed precision model in parallel, but I already state that in the preceding "The global policy is not thread local" section and that is true regardless of where the global policy is set.

File: rfcs/20200929-keras-mixed-precision.md [hunk 3]
```diff
@@ -0,0 +1,1152 @@
+# Keras Mixed Precision API
```
Comment (di-d6) [inline rfcs/20200929-keras-mixed-precision.md:1]: 遵义

File: rfcs/20200929-keras-mixed-precision.md [hunk 4]
```diff
@@ -0,0 +1,1178 @@
...
+
+To use mixed precision in a model, the global policy must be set to "mixed_float16" or "mixed_bfloat16" before the model is constructed. For many models, this is all that is required to use mixed precision.
+
+```python
+# Use mixed precision, for Volta+ GPUs
+tf.keras.mixed_precision.set_global_policy("mixed_float16")
+
+# Use mixed precision, for Cooper Lake CPUs, Ampere GPUs, or Google TPUs
+tf.keras.mixed_precision.set_global_policy("mixed_bfloat16")
+```
```
Comment (lgeiger) [inline rfcs/20200929-keras-mixed-precision.md:204]: The deprecated mixed precision graph rewrite [opted out of using mixed precision on non-optimal hardware](https://github.com/tensorflow/tensorflow/blob/28df4ade506c940f92d7e2cadca5ca4aae26c9ab/tensorflow/core/grappler/optimizers/auto_mixed_precision.cc#L2007-L2009).
I expect users will now need to handle environment detection manually or write custom code to detect the current hardware? I think this might lead to confusion when running in heterogeneous computing environments which requires users to be aware which policy is optimal on which hardware platform.

How do you feel about adding an `auto` policy (alternatively called `optimal` or `mixed`) that would autoselect between either `float32`, `mixed_float16` or `mixed_bfloat16` depending on the hardware support? This would mean that there are no code changes required when switching between CPU only, Volta GPUs or Ampere GPUs systems.

File: rfcs/20200929-keras-mixed-precision.md [hunk 5]
```diff
@@ -0,0 +1,1178 @@
...
+
+LossScaleOptimizer subclasses `tf.keras.optimizers.Optimizer`. For the most part, LossScaleOptimizer simply overrides the base Optimizer methods to delegate to the wrapped optimizer, but adds loss scaling to certain methods.
+
+There are four steps that LossScaleOptimizer must perform in order to implement loss scaling:
+
+1. Multiply the loss by the loss scale.
+2. Divide the gradients by the loss scale.
+3. For a DynamicLossScale, update the loss scale. This means increasing or decreasing the loss scale and updating `num_good_steps` in accordance with the dynamic loss scaling algorithm.
+4. For a DynamicLossScale, skip applying gradients if they are not finite. Gradients are not finite if they have an Inf, -Inf, or NaN value. If any gradient of any replica has a nonfinite value, all gradients across all replicas are skipped for that step. For a FixedLossScale, gradients are unconditionally applied, just like when loss scaling is not used.
```
Comment (lgeiger) [inline rfcs/20200929-keras-mixed-precision.md:693]: What is the benefit of skipping update steps instead of recomputing the gradients with a scaled down loss until the gradients become finite (or until fixed number of retries is reached)? Intuitively this would enable slightly higher performance as the forward pass could be reused, or am I missing something?

**Review by lgeiger — COMMENTED** [review]
@reedwm Thank you very much for the in-depth writeup. I made a few minor comments below, I am currious about what you think.

File: rfcs/20200929-keras-mixed-precision.md [hunk 6]
```diff
@@ -0,0 +1,1178 @@
...
+    ...
+```
+
+The `name` argument determines the compute dtype and the variable dtype. It also determines the loss scale of the policy if the user does not pass in a loss scale to the constructor. The possible values for `name` are:
+
+* Any dtype: The compute and variable dtype is `name`. By default, there is no loss scale. `name` can be a DType, or any value convertible to a DType (including strings).
+* `"mixed_float16"`: The compute dtype is float16. The variable dtype is float32. The default loss scale is "dynamic".
+* `"mixed_bfloat16"`: The compute dtype is bfloat16. The variable dtype is float32. There is no default loss scale, as loss scaling is only useful when float16 is used.
+
+Unlike most TensorFlow functions with a `name` argument, the Policy `name` argument has a semantic impact on the TensorFlow program, and is not just used to uniquely identify an op or layer. The word "name" is chosen for lack of a better word to call the argument.
```
[the remark this answers is not in the document]
Comment (reedwm) [reply rfcs/20200929-keras-mixed-precision.md:100]: The issue is the argument is also exposed as as a property, so if it were called `dtype` instead of `name`, Policy would have a `dtype` field, a `compute_dtype` field, and a `variable_dtype` field. It would be nonobvious that the `dtype` field was mostly unused and only used to determine the `compute_dtype` and `variable_dtype` fields.

File: rfcs/20200929-keras-mixed-precision.md (hunk 5 above)
Comment (reedwm) [reply rfcs/20200929-keras-mixed-precision.md:695]: This is a good idea, but the issue is that we cannot run the loop if the user calls `apply_gradients` instead of `minimize` because `apply_gradients` does not compute the gradients. For example:

```
vars = ...
opt = tf.keras.mixed_precision.LossScaleOptimizer(...)
with tf.GradientTape() as tape:
  loss = get_model_loss()
  grads = tape.gradient(loss, vars)
  opt.apply_gradients(zip(grads, vars))
```

Since the gradients are computed outside the LossScaleOptimizer, LossScaleOptimizer has no way of repeatedly recomputing the gradients.

However, we can do this in `minimize`. I added a paragraph suggesting we should add an option to do this in the future (starting with "Instead of skipping steps when there are NaNs"). The reason this won't be done initially is that `minimize` is rarely used, but I anticipate it will be used more frequently due to #234 allowing users to pass a tensor to `minimize` once TF 2.4 is released.

I think the performance gain will be negligible, due to steps only being skipped 1/2000 of the time on average, but the behavior will be more intuitive. The skipping behavior is confusing because for the first few steps, the variables are not actually updated since the loss scale starts out so high.

File: rfcs/20200929-keras-mixed-precision.md (hunk 4 above)
Comment (reedwm) [reply rfcs/20200929-keras-mixed-precision.md:204]: I strongly considered this earlier. My main concern was that I believed users should be aware of whether they were using float16 or bfloat16. Users need to explicitly use a LossScaleOptimizer if using a custom training loop with float16 (when using `Model.fit`, a LossScaleOptimizer will automatically be used but otherwise the user must explicitly wrap their optimizer). Additionally, the SavedModel and checkpoint format will save the loss scale if mixed_float16 is used, and I don't think the SavedModel/checkpoint format should depend solely on the device that is used.

I added a paragraph summarizing why we don't do this.

Also, the batch size you should run at depends on the device as well, as different GPUs have different amounts of memory, so already users must be somewhat aware of their devices.

In practice, I expect most models supporting multiple devices to have a flag to choose between float32, mixed_float16, and mixed_bfloat16, and that flag is directly passed to `set_global_policy`. Within Google, we do this but also have a per-device per-model config file specifying what flags to pass, so running the GPU config file causes mixed_float16 to be passed to the flag and running the TPU config file causes mxied_bfloat16 to be passed.

File: rfcs/20200929-keras-mixed-precision.md [hunk 7]
```diff
@@ -0,0 +1,1182 @@
...
+* `LossScaleOptimizer` exposes the `inner_optimizer` attribute. [Link](#lossscaleoptimizer-api).
+* `LossScaleOptimizer` delegates `__getattribute__` to the wrapped optimizer if the attribute is a hyperparameter. [Link](#lossscaleoptimizer-__getattribute__-and-__setattr__-delegation).
+* `LossScaleOptimizer` delegates `_setattr__` to the wrapped optimizer if the attribute is a hyperparameter. [Link](#lossscaleoptimizer-__getattribute__-and-__setattr__-delegation).
+* [RFC #234](https://github.com/tensorflow/community/pull/234), titled "Easily Customizable Optimizer.minimize", is implemented. This is not part of the mixed precision API but has some features that make loss scaling easier to use. In particular, it allows a tensor instead of a function to be passed to `Optimizer.minimize`, making `minimize` easier to use. `LossScaleOptimizer.minimize` is recommended over `LossScaleOptimizer.apply_gradients`, as the former automatically scales the loss and gradients while the latter requires the user to manually scale the loss and gradients.
+
+# Alternatives Considered
+
+## Op-based autocasting API
```
Comment (nluehr) [inline rfcs/20200929-keras-mixed-precision.md:972]: Op-based autocasting with down-graph type inheritance is important for providing robust accuracy in mixed precision training. This is particularly true in experimental and research applications. For example, early implementations of batch- and layer-norm used a sequence of low-level operations such as the following.
 
bn(x) = y * (x-mean(x)) / sqrt(var(x)+epsilon) + beta
 
Without computing the mean and variance ops in fp32, training will either diverge or achieve reduced accuracy for many important networks such as InceptionV3, ResNet50, and Xception.
 
Of course, once the utility of an experimental layer is proven, it may be replaced with a “fused” custom kernel that provides optimized performance and handles intermediate precisions internally. However, relatively few layers will merit this effort. Novel model architectures are an important area of innovation, and mixed precision needs to be robust so that researches and early adopters can benefit from the full performance of modern architectures.

File: rfcs/20200929-keras-mixed-precision.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (lgeiger) [reply rfcs/20200929-keras-mixed-precision.md:100]: Thanks for the explanation, that makes sense :+1:

File: rfcs/20200929-keras-mixed-precision.md (hunk 4 above)
Comment (lgeiger) [reply rfcs/20200929-keras-mixed-precision.md:204]: I see your point that users might need to customize other training parameters depending on the device anyway which is indeed very common.

I think the main reason why I was missing this autoselection is that I often end up having model configs which enable mixed precision by default. However, sometimes it can be useful to run a few steps locally on a CPU only machine for debugging which now would require users to remember to disable mixed precision as it would lead to greatly decreased performance on CPU (at least that was the case a few months ago when I last tested this API).
I don't think this is a huge issue since one can easily check for the existance of a GPU in user code and disable mixed precision if needed, but if this becomes a common source of confusion it might be interesting to rethink this behaviour.

File: rfcs/20200929-keras-mixed-precision.md (hunk 5 above)
Comment (lgeiger) [reply rfcs/20200929-keras-mixed-precision.md:695]: Yes, sorry I should've been clearer. I was mainly suggesting this for the use in `keras.Model.fit()` which indeed calls [`.minimize()`](https://github.com/tensorflow/tensorflow/blob/dc843fad65fb4eb1b173a7577d0d932b2af94951/tensorflow/python/keras/engine/training.py#L743). But since for custom training loops the skipping behaviour still needs to be supported anyway, I agree it makes sense to keep the skipping behaviour for now.

**Comment by fchollet** [comment]
We did the internal design review for this API today. Since there were no outstanding issues, it was very quick.

## Notes:

- Reed: Let's present the RFC and go over the changes that happened since we sent it out.
- Before the changes: there were 3 dynamic loss scale classes. They were glorified named tuples
- Now: everything related to loss scaling moved to `LossScaleOptimizer`
  - There's a `dynamic` boolean argument to use dynamic scaling or not
    - if true, you can use initial_scale and dynamic_growth_steps
    - It's a simple API
- Q from Tomer: do you always have to use loss scale optimizer when using mixed precision?
  - Reed: it's automated with compile/fit, but with a custom training loop you need to use it
  - This is well documented

## Status:

Design approved after the minor changes described during the meeting.

**Review by ematejska — APPROVED** [review]
